client/app/pages/queries/utils/archiveQuery.jsx

```diff
@@ -0,0 +1,44 @@
+import { clone, extend } from "lodash";
+import React from "react";
+import Modal from "antd/lib/modal";
+import { Query } from "@/services/query";
+import notification from "@/services/notification";
+
+function confirmArchive() {
+  return new Promise((resolve, reject) => {
+    Modal.confirm({
+      title: "Archive Query",
+      content: (
+        <React.Fragment>
+          <div className="m-b-5">Are you sure you want to archive this query?</div>
+          <div>All alerts and dashboard widgets created with its visualizations will be deleted.</div>
+        </React.Fragment>
+      ),
+      okText: "Archive",
+      okType: "danger",
+      onOk: () => {
+        resolve();
+      },
+      onCancel: () => {
+        reject();
+      },
+      maskClosable: true,
+      autoFocusButton: null,
+    });
+  });
+}
+
+function doArchiveQuery(query) {
+  // Prettier will put `.$promise` before `.catch` on next line :facepalm:
+  // prettier-ignore
+  return Query.delete({ id: query.id }).$promise
+    .then(() => Promise.resolve(extend(clone(query), { is_archived: true, schedule: null })))
```

Comment:
I'd have preferred `Promise.resolve({ ...query, is_archived: true, schedule: null })`, but that works

Reply:
Spread operator will create a plain object and copy properties to it - so query will loose it's prototype with all methods (like `.isNew()`). Therefore I used `clone` + `extend` here